File: cleanos/flutter/lib/shared_widgets_os/relatorio_os.dart
/// relatorio_os.dart — Montagem PURA do RELATÓRIO FINAL da OS ao cliente.
///
/// Porte de `web/src/lib/os/relatorioOS.ts` (buildRelatorioOS) + `calcTotalOS`.
/// Toda a lógica é pura e síncrona: não toca rede nem UI. O caller (tela de
/// execução) materializa o [RelatorioOS] a partir do record PB, e os widgets
/// compartilhados (modal/PDF) consomem o resultado. Classe simples e imutável —
/// sem codegen — para que Painel (Time A) e Profissional (Time B) reusem sem
/// depender de nenhuma feature.
library;

import '../core/models/os_execucao.dart';
import '../core/models/servico.dart';
import 'labels.dart';

/// Pacote pronto para pré-visualizar / virar PDF. Espelha `RelatorioOS` (types.ts).
class RelatorioOS {
  const RelatorioOS({
    required this.osId,
    this.numeroOS,
    required this.clienteNome,
    this.clienteTelefone,
    this.enderecoCompleto,
    this.bairro,
    this.profissionalNome,
    required this.dataHora,
    required this.snapshot,
    required this.adicionais,
    required this.checklist,
    required this.evidencias,
    required this.observacoesVisiveis,
    this.orientacoesPos,
    required this.valorPrincipal,
    required this.valorAdicionais,
    this.descontos,
    required this.valorTotal,
    required this.textoPadrao,
    required this.prazoIntercorrenciaDias,
    this.avaliacaoNota,
    required this.geradoEm,
  });

  final String osId;
  final String? numeroOS;

  final String clienteNome;

  /// 🔒 nunca preenchido pela superfície do PROFISSIONAL (anti-desvio). Existe no
  /// tipo apenas para o Painel poder reusar o mesmo laudo.
  final String? clienteTelefone;
  final String? enderecoCompleto;
  final String? bairro;
  final String? profissionalNome;
  final String dataHora;

  final ServiceSnapshot snapshot;
  final List<ServicoAdicionalOS> adicionais;

  final List<ChecklistExecItem> checklist;
  final List<EvidenciaFoto> evidencias;
  final List<ObservacaoProfissional> observacoesVisiveis;
  final String? orientacoesPos;

  final double valorPrincipal;
  final double valorAdicionais;
  final double? descontos;
  final double valorTotal;

  final String textoPadrao;
  final int prazoIntercorrenciaDias;

  final double? avaliacaoNota;

  /// ISO datetime de quando o relatório foi gerado.
  final String geradoEm;
}

/// Um adicional entra na cobrança (e no relatório do cliente) quando está
/// 'aprovado' ou 'nao_requer'. 'aguardando'/'recusado' não contam. Espelha
/// `isAdicionalCobravel`/`calcTotalOS`.
bool isAdicionalCobravel(ServicoAdicionalOS a) =>
    a.aprovacao == AprovacaoStatus.aprovado ||
    a.aprovacao == AprovacaoStatus.naoRequer;

/// Total da OS: principal + Σ adicionais cobráveis − descontos (nunca negativo).
/// Espelha `calcTotalOS`.
double calcTotalOS(
  double valorPrincipal,
  List<ServicoAdicionalOS> adicionais,
  double? descontos,
) {
  final extras = adicionais
      .where(isAdicionalCobravel)
      .fold<double>(0, (sum, a) => sum + a.valor * a.quantidade);
  final total = valorPrincipal + extras - (descontos ?? 0);
  return total < 0 ? 0 : total;
}

/// Entrada de [buildRelatorioOS]. Campos derivados são calculados na montagem.
class BuildRelatorioOSInput {
  const BuildRelatorioOSInput({
    required this.osId,
    this.numeroOS,
    required this.clienteNome,
    this.clienteTelefone,
    this.enderecoCompleto,
    this.bairro,
    this.profissionalNome,
    required this.dataHora,
    required this.snapshot,
    this.adicionais = const [],
    this.checklist = const [],
    this.evidencias = const [],
    this.observacoes = const [],
    this.descontos,
    this.avaliacaoNota,
    required this.geradoEm,
  });

  final String osId;
  final String? numeroOS;
  final String clienteNome;
  final String? clienteTelefone;
  final String? enderecoCompleto;
  final String? bairro;
  final String? profissionalNome;
  final String dataHora;
  final ServiceSnapshot snapshot;
  final List<ServicoAdicionalOS> adicionais;
  final List<ChecklistExecItem> checklist;
  final List<EvidenciaFoto> evidencias;
  final List<ObservacaoProfissional> observacoes;
  final double? descontos;
  final double? avaliacaoNota;

  /// ISO datetime de "agora" — injetado pelo caller (mantém a função pura).
  final String geradoEm;
}

/// Monta o [RelatorioOS] a partir dos dados da execução. Espelha `buildRelatorioOS`.
RelatorioOS buildRelatorioOS(BuildRelatorioOSInput input) {
  final cobraveis = input.adicionais.where(isAdicionalCobravel).toList();
  final valorPrincipal = input.snapshot.valorBase;
  final valorAdicionais = cobraveis.fold<double>(
    0,
    (sum, a) => sum + a.valor * a.quantidade,
  );
  final valorTotal = calcTotalOS(
    valorPrincipal,
    input.adicionais,
    input.descontos,
  );

  return RelatorioOS(
    osId: input.osId,
    numeroOS: input.numeroOS,
    clienteNome: input.clienteNome,
    clienteTelefone: input.clienteTelefone,
    enderecoCompleto: input.enderecoCompleto,
    bairro: input.bairro,
    profissionalNome: input.profissionalNome,
    dataHora: input.dataHora,
    snapshot: input.snapshot,
    adicionais: cobraveis,
    checklist: input.checklist,
    evidencias: input.evidencias,
    observacoesVisiveis: input.observacoes
        .where((o) => o.visivelCliente)
        .toList(),
    orientacoesPos: input.snapshot.orientacoesPosServico,
    valorPrincipal: valorPrincipal,
    valorAdicionais: valorAdicionais,
    descontos: input.descontos,
    valorTotal: valorTotal,
    textoPadrao: kRelatorioTextoPadrao,
    prazoIntercorrenciaDias: kRelatorioPrazoDias,
    avaliacaoNota: input.avaliacaoNota,
    geradoEm: input.geradoEm,
  );
}

/// Número humano da OS a partir do id (espelha `numeroFromId`).
String numeroFromId(String id) =>
    '#${(id.length <= 6 ? id : id.substring(id.length - 6)).toUpperCase()}';

/// Link de avaliação enviado ao cliente (produção). Espelha `avaliacaoLink`.
String avaliacaoLink(String osId) =>
    'https://cleanox.wenox.com.br/avaliar/${Uri.encodeComponent(osId)}';

File: cleanos/web/src/lib/os/relatorioOS.ts
/**
 * os/relatorioOS.ts — Montagem do RELATÓRIO FINAL da OS ao cliente.
 *
 * Pega os dados crus da execução (snapshot do serviço, adicionais, checklist,
 * evidências e observações do profissional) e produz um {@link RelatorioOS}
 * pronto para pré-visualizar, enviar por WhatsApp ou virar PDF.
 *
 * Toda a lógica é PURA e síncrona — não toca rede nem DOM. As regras financeiras
 * reaproveitam {@link calcTotalOS} para ficar consistentes com o resto da OS.
 *
 * Os dados vêm da OS REAL no PocketBase (service_snapshot, checklist_exec,
 * adicionais e observacoes_prof em `ordens_servico`; evidências em `os_evidencias`
 * com URLs reais). Estas funções permanecem puras: o caller (página de execução /
 * modal) é quem materializa o {@link BuildRelatorioOSInput} a partir do registro PB.
 * O ENVIO em si é server-side (POST /api/cleanos/os/{id}/relatorio, infra uazapi);
 * {@link buildWhatsAppMessage} serve a pré-visualização/cópia no cliente.
 */

import type {
  ChecklistExecItem,
  EvidenciaFoto,
  ObservacaoProfissional,
  RelatorioOS,
  ServiceSnapshot,
  ServicoAdicionalOS,
} from '../servicos/types'
import {
  RELATORIO_PRAZO_DIAS,
  RELATORIO_TEXTO_PADRAO,
  aprovacaoLabel,
  formatTempoMedio,
} from '../servicos/labels'
import { calcTotalOS } from '../servicos/store'
import { formatCurrency, formatDateTime } from '../collections'

/** Entrada de {@link buildRelatorioOS}. Campos derivados são calculados aqui. */
export interface BuildRelatorioOSInput {
  osId: string
  numeroOS?: string
  clienteNome: string
  clienteTelefone?: string
  enderecoCompleto?: string
  bairro?: string
  profissionalNome?: string
  /** ISO datetime do agendamento/execução. */
  dataHora: string
  snapshot: ServiceSnapshot
  adicionais: ServicoAdicionalOS[]
  checklist: ChecklistExecItem[]
  evidencias: EvidenciaFoto[]
  observacoes: ObservacaoProfissional[]
  descontos?: number
  avaliacaoNota?: number
}

/**
 * Um adicional entra na cobrança (e no relatório do cliente) quando está
 * 'aprovado' ou 'nao_requer'. 'aguardando'/'recusado' não contam.
 * Mesma regra de {@link calcTotalOS}.
 */
function isAdicionalCobravel(a: ServicoAdicionalOS): boolean {
  return a.aprovacao === 'aprovado' || a.aprovacao === 'nao_requer'
}

/**
 * Monta o pacote {@link RelatorioOS} a partir dos dados da execução.
 *
 * Regras:
 *  - `observacoesVisiveis` = só observações com `visivelCliente === true`.
 *  - `adicionais` = só os cobráveis (aprovado/nao_requer) — é o que o cliente vê.
 *  - `valorPrincipal` = snapshot.valorBase.
 *  - `valorAdicionais` = Σ (valor × quantidade) dos adicionais cobráveis.
 *  - `valorTotal` = calcTotalOS(principal, adicionais, descontos) (nunca negativo).
 *  - `orientacoesPos` = snapshot.orientacoesPosServico.
 *  - `textoPadrao` / `prazoIntercorrenciaDias` = constantes do módulo.
 *  - `geradoEm` = agora (ISO).
 */
export function buildRelatorioOS(input: BuildRelatorioOSInput): RelatorioOS {
  const adicionaisCobraveis = input.adicionais.filter(isAdicionalCobravel)

  const valorPrincipal = input.snapshot.valorBase
  const valorAdicionais = adicionaisCobraveis.reduce(
    (sum, a) => sum + a.valor * a.quantidade,
    0,
  )
  const valorTotal = calcTotalOS(
    valorPrincipal,
    input.adicionais,
    input.descontos,
  )

  return {
    osId: input.osId,
    numeroOS: input.numeroOS,

    clienteNome: input.clienteNome,
    clienteTelefone: input.clienteTelefone,
    enderecoCompleto: input.enderecoCompleto,
    bairro: input.bairro,
    profissionalNome: input.profissionalNome,
    dataHora: input.dataHora,

    snapshot: input.snapshot,
    adicionais: adicionaisCobraveis,

    checklist: input.checklist,
    evidencias: input.evidencias,
    observacoesVisiveis: input.observacoes.filter((o) => o.visivelCliente),
    orientacoesPos: input.snapshot.orientacoesPosServico,

    valorPrincipal,
    valorAdicionais,
    descontos: input.descontos,
    valorTotal,

    textoPadrao: RELATORIO_TEXTO_PADRAO,
    prazoIntercorrenciaDias: RELATORIO_PRAZO_DIAS,

    avaliacaoNota: input.avaliacaoNota,
    geradoEm: new Date().toISOString(),
  }
}

/**
 * Link de avaliação enviado ao cliente. Aponta para o domínio real de produção
 * (app.cleanox.com.br) na rota /avaliar/{osId}.
 */
export function avaliacaoLink(rel: RelatorioOS): string {
  return `https://app.cleanox.com.br/avaliar/${encodeURIComponent(rel.osId)}`
}

/**
 * Texto formatado para envio por WhatsApp (emojis discretos, markdown leve do
 * WhatsApp com *negrito*). Resume serviço, adicionais, total, checklist
 * executado, orientações pós, prazo de intercorrência e link de avaliação.
 */
export function buildWhatsAppMessage(rel: RelatorioOS): string {
  const lines: string[] = []

  const numero = rel.numeroOS ? ` Nº ${rel.numeroOS}` : ''
  lines.push(`🧼 *Cleanox — Relatório de Serviço${numero}*`)
  lines.push('')
  lines.push(`Olá, ${rel.clienteNome}! Seu serviço foi concluído. ✅`)
  lines.push('Segue o resumo do que foi executado:')
  lines.push('')

  /* Serviço principal */
  lines.push(`🛠️ *Serviço:* ${rel.snapshot.nome}`)
  const tempo = formatTempoMedio(
    rel.snapshot.tempoMedioMin,
    rel.snapshot.tempoMedioLabel,
  )
  if (tempo) lines.push(`⏱️ Tempo médio: ${tempo}`)
  lines.push(`📅 Data: ${formatDateTime(rel.dataHora)}`)
  if (rel.profissionalNome) {
    lines.push(`👤 Profissional: ${rel.profissionalNome}`)
  }
  lines.push('')

  /* Adicionais (só os cobráveis, já filtrados na montagem) */
  if (rel.adicionais.length > 0) {
    lines.push('➕ *Serviços adicionais:*')
    for (const a of rel.adicionais) {
      const qtd = a.quantidade > 1 ? ` (x${a.quantidade})` : ''
      const subtotal = formatCurrency(a.valor * a.quantidade)
      lines.push(`   • ${a.nome}${qtd} — ${subtotal} · ${aprovacaoLabel(a.aprovacao)}`)
    }
    lines.push('')
  }

  /* Resumo financeiro */
  lines.push('💰 *Resumo financeiro:*')
  lines.push(`   Serviço: ${formatCurrency(rel.valorPrincipal)}`)
  if (rel.valorAdicionais > 0) {
    lines.push(`   Adicionais: ${formatCurrency(rel.valorAdicionais)}`)
  }
  if (rel.descontos && rel.descontos > 0) {
    lines.push(`   Descontos: -${formatCurrency(rel.descontos)}`)
  }
  lines.push(`   *Total: ${formatCurrency(rel.valorTotal)}*`)
  lines.push('')

  /* Checklist executado */
  const concluidos = rel.checklist.filter((c) => c.status === 'concluido')
  const pendentes = rel.checklist.filter((c) => c.status !== 'concluido')
  if (rel.checklist.length > 0) {
    lines.push(
      `📋 *Checklist executado* (${concluidos.length}/${rel.checklist.length}):`,
    )
    for (const item of concluidos) {
      lines.push(`   ✓ ${item.titulo}`)
    }
    for (const item of pendentes) {
      lines.push(`   ◻️ ${item.titulo} (pendente)`)
    }
    lines.push('')
  }

  /* Observações visíveis ao cliente */
  if (rel.observacoesVisiveis.length > 0) {
    lines.push('📝 *Observações:*')
    for (const obs of rel.observacoesVisiveis) {
      lines.push(`   • ${obs.texto}`)
    }
    lines.push('')
  }

  /* Orientações pós-serviço */
  if (rel.orientacoesPos && rel.orientacoesPos.trim()) {
    lines.push('🧴 *Orientações pós-serviço:*')
    lines.push(`   ${rel.orientacoesPos.trim()}`)
    lines.push('')
  }

  /* Prazo de intercorrência */
  lines.push(
    `⏳ Você tem até *${rel.prazoIntercorrenciaDias} dias* para relatar ` +
      'qualquer falha ou intercorrência. Conte com a gente!',
  )
  lines.push('')

  /* Avaliação */
  lines.push(`⭐ Avalie nosso atendimento: ${avaliacaoLink(rel)}`)
  lines.push('')
  lines.push('Obrigado por escolher a Cleanox! 💙')

  return lines.join('\n')
}
